Implement fast_sigmoid C ++ extension and Python Binding APIs

hieplcc opened the pull request:
**Implement fast_sigmoid C ++ extension and Python Binding APIs**
1) fast_sigmoid C++ implementation: 
`- extension-cpp/extension_cpp/csrc/include/extension_cpp/fast_sigmoid.h`
`- extension-cpp/extension_cpp/csrc/fast_sigmoid.cpp`
2) fast_sigmoid C++ testing:
`- extension-cpp/test/fast_sigmoid/fast_sigmoid_cpp_test.cpp`
3) CMake file for building fast_sigmoid C++ test:
`- extension-cpp/CMakeLists.txt`
`- extension-cpp/BUILD.md`
4) fast_sigmoid Python Binding:
`- extension-cpp/extension_cpp/ops.py`
5) fast_sigmoid Python testing:
`- extension-cpp/test/fast_sigmoid/test_fast_sigmoid.py`

hieplcc committed:
Implement fast_sigmoid C ++ extension

hieplcc committed:
Addinng simple test for fast_sigmoid C++ extension

hieplcc committed:
Adding Python Binding APIs for fast_sigmoid

hieplcc committed:
Adding thread-safe mechanism for look-up table cache

hieplcc committed:
Adding option of using LRU cache for the look-up table

hieplcc committed:
Adding more test cases for C++ and Python Binding

hieplcc committed:
Optimize branch prediction

loctxmoreh commented on test/fast_sigmoid/test_fast_sigmoid.py:
* You should test numerical correctness using `all_close` method 
* Use `opcheck` from `torch.testing._internal.optest` for a complete check

hieplcc replied:
Dear @loctxmoreh 
I've just pushed a commit to Improve fast_sigmoid python test using opcheck and torch.allclose.
I also fixed a potential out of bounds bug in the interpolation logic.
Please help me to review.
Thanks.

loctxmoreh reviewed (commented):
Minor comments about testing only. Otherwise look good

hieplcc committed:
Improve fast_sigmoid python test and fix edge cases:

- Add opcheck-based test, use torch.allclose for numerical correctness

- Prevent out-of-bounds indexing in interpolation logic

- Fix typo in fast_sigmoid_cpu implementation

loctxmoreh commented on extension_cpp/csrc/fast_sigmoid.cpp:
```diff
@@ -114,7 +114,11 @@ at::Tensor fast_sigmoid_cpu(const at::Tensor& input, double min_val, double max_
+
+            // Make sure that lower is less than num_entries - 1. Otherwise, upper (lower + 1) will be out of bounds
+            // (num_entries - 2 >= 0 due to num_entries > 1 check)
+            int64_t lower = std::min(idx, num_entries - 2);
```
Hmm good catch. This would be a pretty subtle bug. 
Can you write an unittest (in either python or cpp) that will trigger the bug _before_ you come up with this fix? 
Particularly, you should be able to find a `x` that satisfies both `x < max_val` and `floor(idx_f) == num_entries - 1`

hieplcc replied:
Dear @loctxmoreh 
Initially, I'm concerned about the floating point precision, that can lead to incorrect comparison like this example:
https://www.onlinegdb.com/t-Lwf3l4I
I reviewed again, and  if `min_val < x < max_val`, then `idx` is guaranteed to be in the range `[0, num_entries - 2]`.
So, its safely to remove the out-of-bound check. Actually, we can defensively  put the clamping check here, but it would introduce minor overheads (of `std::min`, `std::max`). 
Furthermore, I added the code to cast the `min_val `and `max_val` to float before processing. This helps to have consistent precision behavior, and reduce the cost (time) of casting between doubles and floats multiple times.
Please help to re my commit https://github.com/hieplcc/extension-cpp/pull/1/commits/0951dadccbdc81105d1a2e9175ea55b59559a216

loctxmoreh replied:
> I reviewed again, and if min_val < x < max_val, then idx is guaranteed to be in the range [0, num_entries - 2].

Actually I disagree. Algebraically, it is correct that if min_val < x < max_val then idx is guaranteed to be in valid range.  However due to rounding error, the bug might still happens. 

> find a x that satisfies both x < max_val and floor(idx_f) == num_entries - 1

I wrote a small program to find `x` that trigger the bug, given `min_val = -10.0`, `max_val = 10.0` and `num_entries = 1000`. Eventually I found this value: 
```
Found x = 9.9999990463, idx_f = 999.0000000000, floor(idx_f) = 999.0000000000, hex(x) = 0x411FFFFF
```

Pls check to see if the found `x` value actually trigger the out-of-bound error, given your current implementation.

hieplcc replied:
Thank you for pointing out the case. It is really interesting.
I used the `x` value you provided in my implementation and the `idx` is 999.0, which leads to out-of-bounds.
I tried to find the way to find a value `x` like you, and come up with this approach: https://onlinegdb.com/6Qp4leREa in which I find the nearest value of  `max_val`.
I also tried to find the nearest value of `min_val` to see if there is a `x` value that make `idx` equal to `-1` but I have not found it yet.
However I think there might be a `x` value like that, so in the fix, I clamp the value of `idx` in the range of `[0, num_entries - 2]`.
Here is my fix commit https://github.com/hieplcc/extension-cpp/pull/1/commits/59ef092939217c0fb68672d93da520ec9bdeb872

loctxmoreh commented on test/fast_sigmoid/fast_sigmoid_cpp_test.cpp:
My original intention with this test program is for it to be a standalone client that used `lib_fast_sigmoid` as a dependency. So the build system now should be divided into at least three part
* `extension_cpp/csrc/CMakeLists.txt`: build the library 
* `examples/CMakeLists.txt` and `fast_sigmoid_cpp_test.cpp`: build the client program  
* root `CMakeLists.txt`: delegates to the other two Cmake files 

Furthermore, you can try to package & distribute `lib_fast_sigmoid` using CPack. This is optional, tho

hieplcc replied:
I just pushed a new commit for separating `CMakeLists`.
I also package the `fast_sigmoid` library and test it with an `external_app_demo` using `find_package`.
Please help me to review https://github.com/hieplcc/extension-cpp/pull/1/commits/3e52d8199ef03656ef98e69fce4a0111f1b2d736

loctxmoreh reviewed (commented):
These would be the last two request from me:
* Write test that trigger the out-of-bound index 
* Separate the cpp test program as a client using `lib_fast_sigmoid` as dependency

hieplcc committed:
Remove unnessary out of bounds check. Cast min, max values to float before process

hieplcc committed:
Adding CMake build for fast_sigmoid library

loctxmoreh commented on BUILD.md:
````diff
@@ -0,0 +1,58 @@
+
+```bash
+# From inside the build directory
+make install DESTDIR=../install_demo
````
```suggestion
make install DESTDIR=/path/to/venv
```
Suggestion: since we're developing in both python and cpp, so the existing python virtual env is a good candidate for installing stuff into

loctxmoreh commented on CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+include(CPack)
+# General CPack metadata
+set(CPACK_PACKAGE_NAME "fast_sigmoid")
+set(CPACK_PACKAGE_VERSION ${PROJECT_VERSION})
```
If you intend to support packaging using CPack, pls add the guide to `BUILD.md`
(as i mentioned earlier, this is optional)

loctxmoreh commented on extension_cpp/csrc/CMakeLists.txt:
```diff
@@ -0,0 +1,96 @@
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${TORCH_CXX_FLAGS}")
+
+## Add include directories (for build only)
+include_directories(include ../../utils)
```
This is redundant, due to a call to `target_include_directories` later

loctxmoreh commented on external_app_demo/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+list(APPEND CMAKE_PREFIX_PATH "${TORCH_PREFIX_PATH}")
+
+# Add fast_sigmoid install path
+list(APPEND CMAKE_PREFIX_PATH "../install_demo/usr/local")
```
`../install_demo` is hard-coded here. I prefer no to hard-code it. 
From a POV of a library developer, pls find a sensible way to specify the prefix path

loctxmoreh reviewed (commented):
Minor comments about packaging. Also, 

> Write test that trigger the out-of-bound index

I put in some works and have progress in this topic. I'd like to continue this discussion.

hieplcc committed:
Fix edge cases on floating rounding issues

hieplcc committed:
Improve CMake configurations, add CPack option and update BUILD.md

hieplcc commented:
Dear @loctxmoreh 
I pushed the commit https://github.com/hieplcc/extension-cpp/pull/1/commits/e91114a7900ce705f4ee7e3c366fff5eef6927c1 to fix issues related to building package:
- Improve `CMake` configuration for using `fast_sigmoid` lib.
- Add option for using `CPack`.
- Update the `BUILD.md` for building lib using `CPack`.

loctxmoreh reviewed (approved):
(Sorry for the late review. I've been really busy for the last two days :disappointed: )

LGTM. 

I've finished all my reviews on all PRs. Thank you for your hard works. You'll be informed of the result soon

hieplcc commented:
Dear @loctxmoreh 
Thank you for your detail review.
I’ve learned a lot through this task.
